src/cryptonote_basic/cryptonote_format_utils.cpp

```diff
@@ -857,6 +859,27 @@ namespace cryptonote
     return outputs_amount;
   }
   //---------------------------------------------------------------
+  bool get_output_public_key(const cryptonote::tx_out& out, crypto::public_key &output_public_key)
+  {
+    // before HF_VERSION_VIEW_TAGS, outputs with public keys are of type txout_to_key
+    // after HF_VERSION_VIEW_TAGS, outputs with public keys are of type txout_to_tagged_key
+    if (out.target.type() != typeid(txout_to_key) && out.target.type() != typeid(txout_to_tagged_key))
+      return false;
+
+    output_public_key = out.target.type() == typeid(txout_to_key)
+      ? boost::get< txout_to_key >(out.target).key
+      : boost::get< txout_to_tagged_key >(out.target).key;
+
+    return true;
+  }
+  //---------------------------------------------------------------
+  const crypto::view_tag *get_output_view_tag(const cryptonote::tx_out& out)
```

Comment:
Is it necessary to return a pointer here? View tags are simple chars, so copies are very cheap.

Reply:
The more I sit with this, the less safe it feels. @selsta 's perf tests (that were passing on my machine) [are failing](https://github.com/monero-project/monero/pull/8061#issuecomment-969508340) because of UB I introduced from passing a pointer to a locally scoped view tag. I'd rather not pass around a pointer, just in case a situation like that arises. I was following this pattern for commitments in `add_output`:

https://github.com/monero-project/monero/blob/6d94953be259c6bf4a1407ec552a570a66b6179d/src/blockchain_db/blockchain_db.cpp#L245

But that's isolated to a single function call, rather than spread out across various locations in the code.

Thinking I'll instead return a `std::pair<bool, crytpo::view_tag>`, and the bool indicates whether or not the view tag is usable. Seems a little heavy, but safer + more explicit about the extra information tied to the view tag